File: src/utils/combat.ts
// src/utils/combat.ts
import type { ActionType, Entite } from '../types';

// Plafond de combo imposé aux monstres tant que le joueur n'a achevé aucune run : enchaîner cinq
// fois la même action est le pic de puissance d'une créature, et un débutant ne sait pas encore
// que ça se voit venir dans ses cases d'action. La borne saute dès la 2e ascension.
export const LIMITE_COMBO_PREMIERE_RUN = 4;

export const SYMBOLES: Record<ActionType, string> = { 'A': '⚔️', 'P': '🎯', 'D': '🛡️', 'E': '💨' };

// Le Feu et le Poison ne s'AJOUTENT pas à une action : ils la remplacent (l'Attaque cesse de blesser
// sur le coup, la Précise aussi). Garder ⚔️/🎯 laisserait croire à des dégâts immédiats qui n'ont
// plus lieu — d'où le changement d'icône partout où l'action est représentée. Miroir de `symbole()`
// côté moteur (lib.rs), qui fait de même dans le journal.
export function symbolePour(action: ActionType, entite?: Entite): string {
    if (action === 'A' && entite?.multiplicateurBrulure) return '🔥';
    if (action === 'P' && entite?.multiplicateurPoison) return '🧪';
    return SYMBOLES[action];
}

// ⚠️ L'armure regénérée en début de tour (« Pelage d'Acier », passifs de boss) n'est créditée qu'au
// lancement de la résolution : pendant que le joueur programme, `armure` vaut encore 0 alors que la
// cible en portera à coup sûr quand le coup partira. On l'anticipe, sinon la stat annonce des dégâts
// simples pour un coup qui sera amplifié.
export function cibleSeraArmee(cible?: Entite): boolean {
    if (!cible) return false;
    return cible.armure > 0 || (cible.regenArmureTour ?? 0) > 0;
}

// Pacte de la Foudre : le coup entier est amplifié tant que la CIBLE porte de l'armure. C'est la
// seule contribution qui dépende de l'adversaire, d'où le paramètre `cible` partout en aval — sans
// lui, la stat resterait muette sur l'Étage de la Foudre alors que les dégâts y sont jusqu'à
// triplés. Miroir de `appliquer_foudre()` côté moteur.
function appliquerFoudre(valeur: number, entite: Entite, cible?: Entite): number {
    if (!entite.multiplicateurDegatsSiArmure || !cibleSeraArmee(cible)) return valeur;
    return Math.round(valeur * entite.multiplicateurDegatsSiArmure);
}

// Valeur d'Attaque affichée dans les stats : contrairement aux pactes à bonus plat (ex: Pacte de
// la Vie qui augmente directement pvMax), le Pacte de la Puissance Brute n'agit qu'au moment du
// calcul de dégâts côté moteur Rust — sans ceci, le joueur ne verrait jamais son ⚔️ bouger à
// l'écran alors que le bonus est bien actif. Même formule d'arrondi que le moteur (get_valeur_action).
//
// `actionsEnAttente` (optionnel) permet en plus un aperçu LIVE du bonus temporaire de la Synergie
// Guerrier pendant que le joueur programme son tour : chaque Défense déjà posée dans la file
// ajoute +2 aux dégâts de base affichés, avant même de valider le tour — pour que le bonus
// "temporaire" (remis à zéro chaque tour côté moteur) reste visible plutôt qu'invisible.
export function calculerAttaqueAffichee(entite: Entite, actionsEnAttente: ActionType[] = [], cible?: Entite): number {
    let base = entite.baseA;
    if (entite.synergieActive === 'Guerrier') {
        const nbDefenses = actionsEnAttente.filter(a => a === 'D').length;
        base += 2 * nbDefenses;
    }
    if (entite.bonusDegatsAttaquePourcentage) {
        base = Math.round(base * (1 + entite.bonusDegatsAttaquePourcentage / 100));
    }
    // Pacte du Feu : l'Attaque ne blesse plus, elle pose une brûlure valant une PART de ses dégâts.
    // C'est cette dose que le joueur doit lire sous l'icône 🔥, pas les dégâts qu'il n'infligera pas.
    // ⚠️ La Foudre n'entre dans la brûlure QUE via la Synergie Élémentaire (cf. convertir_en_brulure).
    if (entite.multiplicateurBrulure) {
        const valeur = entite.synergieActive === 'Elementaire' ? appliquerFoudre(base, entite, cible) : base;
        return Math.round(valeur * entite.multiplicateurBrulure);
    }

    return appliquerFoudre(base, entite, cible);
}

// Même principe pour la Précise : le Pacte de l'Ombre (I/II) double ses dégâts (degatsPrecisDoubles)
// et, avec la Synergie Assassin, elle hérite aussi du bonus % du Pacte de la Puissance Brute —
// deux effets "toujours actifs" (contrairement au Pacte du Combo ou du Temps, conditionnés au
// combo/à la position dans le tour) qui doivent donc être visibles directement sur 🎯.
// Même ordre d'opérations que le moteur : bonus % (get_valeur_action) puis doublage (calculer_degats).
export function calculerPreciseAffichee(entite: Entite, cible?: Entite): number {
    let valeur = entite.baseP;
    if (entite.synergieActive === 'Assassin' && entite.bonusDegatsAttaquePourcentage) {
        valeur = Math.round(valeur * (1 + entite.bonusDegatsAttaquePourcentage / 100));
    }

    // Pacte du Poison : la dose se calcule sur la valeur BRUTE de l'action. Ni la Foudre ni le
    // doublage de l'Ombre n'y entrent — ils vivent dans `calculer_degats`, dont une action convertie
    // ressort à zéro (cf. convertir_en_poison).
    if (entite.multiplicateurPoison) return Math.round(valeur * entite.multiplicateurPoison);

    // Ordre imposé par le moteur : la Foudre amplifie, PUIS l'Ombre double.
    valeur = appliquerFoudre(valeur, entite, cible);
    if (entite.degatsPrecisDoubles) valeur *= 2;
    return valeur;
}

// Miroir de paliers_esquive_effectifs() + chance_esquive() côté moteur (combat.rs) : le Pacte du
// Combo amplifie l'écart entre paliers consécutifs à partir du 2e (jamais le 1er, qui n'est jamais
// un combo), puis les Bénédictions du Chat décalent le tout d'un bonus plat ("Grâce Féline",
// visible dès le palier 0) ou de la réduction imposée par l'adversaire ("Regard Hypnotique").
// Sans ceci l'affichage du % d'esquive resterait figé sur les paliers de base, comme pour ⚔️
// avant calculerAttaqueAffichee.
export function calculerPaliersEsquiveAffiches(entite: Entite, reductionAdverse = 0): number[] {
    const base = entite.paliersEsquive;
    const mult = entite.comboMultiplicateur ?? 1;
    const effectifs = [base[0] ?? 0, 0, 0, 0];
    for (let i = 1; i < 4; i++) {
        const precedent = base[i - 1] ?? 0;
        const actuel = base[i] ?? precedent;
        const delta = actuel - precedent;
        const deltaAjuste = i > 1 ? Math.round(delta * mult) : delta;
        effectifs[i] = Math.min(100, effectifs[i - 1] + deltaAjuste);
    }

    const decalage = (entite.bonusEsquiveFlat ?? 0) - reductionAdverse;
    if (decalage === 0) return effectifs;
    return effectifs.map(palier => Math.min(100, Math.max(0, palier + decalage)));
}

export interface CreneauxFroid {
    gelesJoueur: number[];
    gelesMonstre: number[];
    joueurDabord: number[];
    monstreDabord: number[];
    // Créneaux où les deux camps portaient le dérèglement : il s'y neutralise. Purement informatif
    // (jamais transmis au moteur), mais indispensable — sans repère, le joueur qui porte le Pacte du
    // Froid à l'étage du Froid croit simplement que son Pacte ne fonctionne pas.
    annules: number[];
}

export const AUCUN_CRENEAU_FROID: CreneauxFroid = { gelesJoueur: [], gelesMonstre: [], joueurDabord: [], monstreDabord: [], annules: [] };

function tirerCreneaux(nombre: number): number[] {
    if (nombre <= 0) return [];
    return melangerIndices().slice(0, Math.min(nombre, 5));
}

function melangerIndices(): number[] {
    const indices = [0, 1, 2, 3, 4];
    for (let i = indices.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [indices[i], indices[j]] = [indices[j], indices[i]];
    }
    return indices;
}

// Étage du Froid : quels créneaux du tour sont déréglés (résolus avant l'adversaire) et lesquels
// sont gelés. ⚠️ Tiré ICI, au début du tour, et non dans le moteur : le joueur doit voir les cases
// concernées PENDANT qu'il programme ses actions, sinon l'information arrive trop tard pour être
// jouable. Le résultat est passé tel quel à `jouer_tour`.
// Un pouvoir porté des DEUX côtés s'annule : seul l'écart compte. La part neutralisée est quand
// même tirée et rendue dans `annules`, pour pouvoir montrer au joueur OÙ son dérèglement a été
// contré — un tableau vide, lui, ressemblerait à un Pacte inopérant.
export function tirerCreneauxFroid(joueur: Entite, monstre: Entite): CreneauxFroid {
    const inversionsJ = joueur.actionsResolutionInversee ?? 0;
    const inversionsM = monstre.actionsResolutionInversee ?? 0;

    // Un seul mélange pour les deux lots : un créneau ne peut pas être à la fois déréglé et neutralisé.
    const indices = melangerIndices();
    const dabord = indices.slice(0, Math.min(Math.abs(inversionsJ - inversionsM), 5));
    const annules = indices.slice(dabord.length, Math.min(dabord.length + Math.min(inversionsJ, inversionsM), 5));

    return {
        gelesJoueur: tirerCreneaux(monstre.actionsGelees ?? 0),
        gelesMonstre: tirerCreneaux(joueur.actionsGelees ?? 0),
        joueurDabord: inversionsJ > inversionsM ? dabord : [],
        monstreDabord: inversionsM > inversionsJ ? dabord : [],
        annules,
    };
}

// `poisonSurCible` : dose de poison déjà installée sur l'adversaire de cette créature. Elle seule
// décide du droit de temporiser (voir peutTemporiserSiPoisonDepasse).
export function genererActionsMonstre(monstre: Entite, poisonSurCible: number = 0): ActionType[] {
    const possibilites = monstre.actionsPossibles;
    const actions: ActionType[] = [];
    const chancePoursuiteCombo = monstre.chanceCombo !== undefined ? monstre.chanceCombo : 20;

    for (let i = 0; i < 5; i++) {
        let choix: ActionType;
        if (i === 0) {
            choix = possibilites[Math.floor(Math.random() * possibilites.length)];
        } else {
            // Bonus spécifique : après une Défense, certains boss (Pacte de l'Armure) ont une
            // chance accrue d'enchaîner une nouvelle Défense, pour finir le tour avec plus
            // d'armure et déclencher plus souvent leurs dégâts de fin de tour.
            let chanceContinuer = chancePoursuiteCombo;
            if (actions[i - 1] === 'D' && monstre.chanceSuiteDefense) {
                chanceContinuer += monstre.chanceSuiteDefense;
            }

            const jetCombo = Math.random() * 100;
            if (jetCombo < chanceContinuer) choix = actions[i - 1];
            else choix = possibilites[Math.floor(Math.random() * possibilites.length)];
        }
        actions.push(choix);
    }

    // Un tour entier passé à se défendre et esquiver n'a de sens que pour une créature dont le
    // poison déjà injecté fait le travail à sa place. Partout ailleurs, c'est un tour offert au
    // joueur : on force alors au moins une action offensive.
    const peutTemporiser = monstre.peutTemporiserSiPoisonDepasse !== undefined
        && poisonSurCible > monstre.peutTemporiserSiPoisonDepasse;
    const offensives = possibilites.filter(a => a === 'A' || a === 'P');
    if (!peutTemporiser && offensives.length > 0 && !actions.some(a => a === 'A' || a === 'P')) {
        actions[Math.floor(Math.random() * 5)] = offensives[Math.floor(Math.random() * offensives.length)];
    }

    return actions;
}

// Tire, parmi les 5 emplacements d'actions du monstre, les indices qui seront révélés au joueur
// (les autres restent masqués derrière un '❓'). `undefined` ou >= 5 signifie "tout est visible".
export function genererIndicesVisibles(nombreVisible?: number): number[] {
    const tous = [0, 1, 2, 3, 4];
    if (nombreVisible === undefined || nombreVisible >= 5) return tous;

    const melange = [...tous];
    for (let i = melange.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [melange[i], melange[j]] = [melange[j], melange[i]];
    }
    return melange.slice(0, Math.max(0, nombreVisible));
}

// Remplace toute action qui prolongerait un combo au-delà de `limiteCombo` par une alternative
// aléatoire, pour respecter le Pacte de la Fluidité imposé par l'adversaire.
// ⚠️ L'alternative se tire dans la panoplie de la CRÉATURE (`actionsPossibles`), jamais dans la
// liste complète des 4 actions : sinon briser le combo d'un Gardien lui met dans les mains une
// action qu'il ne possède pas — une Précise chez le Brasier Vorace, une Attaque chez la Sève Noire.
export function corrigerActionsPourLimiteCombo(
    actions: ActionType[],
    limiteCombo: number,
    actionsPossibles: ActionType[],
): ActionType[] {
    if (limiteCombo >= 5) return actions;

    const corrigees: ActionType[] = [];
    for (const action of actions) {
        let choix = action;
        let actionsSuite = 0;
        for (let j = corrigees.length - 1; j >= 0; j--) {
            if (corrigees[j] === choix) actionsSuite++; else break;
        }

        if (actionsSuite >= limiteCombo) {
            const alternatives = actionsPossibles.filter(a => a !== choix);
            // Une créature réduite à une seule action ne peut pas briser son propre combo : on la
            // laisse le prolonger plutôt que de lui inventer un geste.
            if (alternatives.length > 0) {
                choix = alternatives[Math.floor(Math.random() * alternatives.length)];
            }
        }
        corrigees.push(choix);
    }
    return corrigees;
}

File: moteur_wasm/src/combat.rs
use crate::entite::{Entite, ActionType, Synergie};
use rand::Rng;

pub struct ResultatDegats {
    pub dmg_arm: i32,
    pub dmg_pv: i32,
    pub esquive: bool,
    // Puissance brute de l'attaque quand elle est esquivée (0 sinon) — sert à afficher les
    // statistiques "dégâts esquivés" en fin de run.
    pub degats_evites: i32,
    // Chance d'esquive qui s'appliquait à ce coup, réussie ou non. Affichée dans le journal : sans
    // elle, un enchaînement d'esquives ressemble à de la malchance pure au lieu d'un palier atteint.
    pub chance_esquive: i32,
    // Dose de brûlure / de poison réellement posée par ce coup (voir convertir_en_* dans lib.rs).
    // Une action convertie n'inflige aucun dégât direct : sans ces deux champs, le journal n'aurait
    // strictement rien à dire d'un coup qui a pourtant porté.
    pub brulure_posee: i32,
    pub poison_pose: i32,
    // Le multiplicateur de la Foudre a bien joué sur ce coup (la cible portait de l'armure au moment
    // précis de la frappe). Impossible à recalculer depuis le journal : l'armure de la cible bouge
    // à l'intérieur même du créneau, selon l'ordre de résolution.
    pub foudre_appliquee: bool,
    // Valeur RÉELLEMENT portée par le coup, une fois tout appliqué (Foudre, doublage de la Précise,
    // conversion en brûlure/poison). Le journal affiche ça et non la valeur brute du combo : sur
    // l'Étage de la Foudre, la seconde ment d'un facteur 1,5 à 3.
    pub valeur_appliquee: i32,
}

// `fusion_ap` : Synergie Assassin ("Danse des Lames") — Attaque et Précise comptent comme la même
// action pour la jauge de combo (A-A-P-P-P = Combo x5), sans quoi alterner les deux briserait le
// combo comme n'importe quel changement d'action.
pub fn gerer_combo(combo_type: &mut Option<ActionType>, combo_count: &mut i32, action: &ActionType, fusion_ap: bool) {
    if *action == ActionType::E {
        *combo_type = None;
        *combo_count = 0;
        return;
    }
    let normalise = if fusion_ap && (*action == ActionType::A || *action == ActionType::P) { ActionType::A } else { action.clone() };
    if Some(normalise.clone()) == *combo_type {
        *combo_count += 1;
    } else {
        *combo_type = Some(normalise);
        *combo_count = 1;
    }
}

// Pacte du Combo appliqué à l'Esquive : NE fait PAS sauter un palier supplémentaire (le niveau
// d'esquive doit rester un décompte fidèle du nombre d'actions Esquive enchaînées, pour que la
// décroissance après une action différente retombe au bon palier). À la place, le multiplicateur
// amplifie l'écart entre paliers consécutifs (à partir du 2e palier, le 1er n'étant jamais un
// combo) : ex. paliers [0,50,75,100] avec x1.5 -> [0,50,88,100] (50 + round(25*1.5) = 88, capé à 100).
fn paliers_esquive_effectifs(entite: &Entite) -> [i32; 4] {
    let base = &entite.paliers_esquive;
    let mult = entite.combo_multiplicateur.unwrap_or(1.0);
    let mut effectifs = [0i32; 4];
    effectifs[0] = base.first().copied().unwrap_or(0);
    for i in 1..4 {
        let precedent = base.get(i - 1).copied().unwrap_or(0);
        let actuel = base.get(i).copied().unwrap_or(precedent);
        let delta = (actuel - precedent) as f32;
        let delta_ajuste = if i > 1 { (delta * mult).round() as i32 } else { delta as i32 };
        effectifs[i] = std::cmp::min(100, effectifs[i - 1] + delta_ajuste);
    }
    effectifs
}

pub fn get_valeur_action(action: &ActionType, count: i32, entite: &Entite) -> i32 {
    let mut multiplier = 1.0;

    // Lecture propre du multiplicateur si on est en situation de combo
    if count > 1 {
        multiplier = entite.combo_multiplicateur.unwrap_or(1.0);
    }

    // Synergie Assassin ("Danse des Lames") : la Précise bénéficie aussi des bonus du Pacte de la
    // Puissance Brute, normalement réservés à l'Attaque.
    let synergie_assassin = entite.synergie_active == Some(Synergie::Assassin);

    // Le match affecte directement la valeur de base, sans variable inutile
    let base_val = match action {
        ActionType::A => {
            // Pacte de la Puissance Brute II : +N dégâts supplémentaires par palier de combo,
            // en plus du bonus de combo standard (+5/palier).
            let bonus_combo_pacte = entite.bonus_combo_attaque_palier.unwrap_or(0) * (count - 1);
            entite.base_a + (count - 1) * 5 + bonus_combo_pacte
        },
        ActionType::P => {
            let bonus_combo_pacte = if synergie_assassin { entite.bonus_combo_attaque_palier.unwrap_or(0) * (count - 1) } else { 0 };
            entite.base_p + (count - 1) * 2 + bonus_combo_pacte
        },
        ActionType::D => entite.base_d + (count - 1) * 5,
        ActionType::E => 0,
    };

    // .round() plutôt qu'un simple `as i32` : un cast tronque toujours vers zéro (16.5 -> 16),
    // alors qu'un bonus de dégâts doit arrondir normalement (16.5 -> 17).
    let valeur = (base_val as f32 * multiplier).round() as i32;

    // Pacte de la Puissance Brute (I/II) : +10%/+20% dégâts sur l'Attaque, appliqué après le combo.
    // (et aussi sur la Précise avec la Synergie Assassin, voir plus haut)
    if *action == ActionType::A || (*action == ActionType::P && synergie_assassin) {
        if let Some(pct) = entite.bonus_degats_attaque_pourcentage {
            return (valeur as f32 * (1.0 + pct as f32 / 100.0)).round() as i32;
        }
    }

    valeur
}

// Multiplicateur d'un Coup Critique de la Bénédiction "Griffe Acérée" (150% des dégâts).
pub const MULTIPLICATEUR_CRITIQUE: f32 = 1.5;

// Chance d'esquive réellement appliquée au défenseur, tous modificateurs compris : le palier
// courant (déjà amplifié par le Pacte du Combo), le bonus plat de la Bénédiction "Grâce Féline"
// (actif même au palier 0, là où les paliers exigent d'avoir joué Esquive) et la réduction imposée
// par l'attaquant ("Regard Hypnotique"). Neutraliser l'esquive (Pacte de l'Ombre II, Le Vent
// Mortel) court-circuite tout le reste.
pub fn chance_esquive(attaquant: &Entite, defenseur: &Entite) -> i32 {
    if attaquant.bloque_esquive_opposant { return 0; }

    let palier = paliers_esquive_effectifs(defenseur)[defenseur.niv_esquive.min(3)];
    let brut = palier
        + defenseur.bonus_esquive_flat.unwrap_or(0)
        - attaquant.reduction_esquive_opposant.unwrap_or(0);
    brut.clamp(0, 100)
}

// Bénédiction "Griffe Acérée" : chance fixe qu'une action offensive frappe en critique. Renvoie la
// valeur éventuellement amplifiée et un drapeau, pour que l'appelant puisse l'annoncer dans le log.
pub fn tenter_critique(valeur: i32, action: &ActionType, attaquant: &Entite) -> (i32, bool) {
    if *action != ActionType::A && *action != ActionType::P { return (valeur, false); }

    let chance = attaquant.chance_critique.unwrap_or(0);
    if chance <= 0 || rand::thread_rng().gen_range(1..=100) > chance { return (valeur, false); }

    ((valeur as f32 * MULTIPLICATEUR_CRITIQUE).round() as i32, true)
}

// Étage de la Foudre : tant que la cible porte de l'armure, le coup entier est amplifié — Précise
// comprise, qui conserve par ailleurs son contournement d'armure. Se défendre contre un porteur de
// ce pouvoir est donc un piège.
// Exposé à part de `calculer_degats` : la Synergie Élémentaire en fait aussi profiter la brûlure,
// qui ne passe pas par le calcul de dégâts normal (voir convertir_en_brulure dans lib.rs).
pub fn appliquer_foudre(val_atk: i32, attaquant: &Entite, defenseur: &Entite) -> i32 {
    match attaquant.multiplicateur_degats_si_armure {
        Some(mult) if defenseur.armure > 0 => (val_atk as f32 * mult).round() as i32,
        _ => val_atk,
    }
}

// L'attaquant est passé en entier (plutôt que ses drapeaux un par un) : c'est lui qui porte à la
// fois le doublage de la Précise, la neutralisation de l'esquive et la réduction d'esquive adverse.
pub fn calculer_degats(action_atk: &ActionType, val_atk: i32, attaquant: &Entite, defenseur: &Entite) -> ResultatDegats {
    let neutre = ResultatDegats { dmg_arm: 0, dmg_pv: 0, esquive: false, degats_evites: 0, chance_esquive: 0, brulure_posee: 0, poison_pose: 0, foudre_appliquee: false, valeur_appliquee: 0 };
    if *action_atk != ActionType::A && *action_atk != ActionType::P {
        return neutre;
    }

    let degats_precis_doubles = attaquant.degats_precis_doubles;

    let val_amplifiee = appliquer_foudre(val_atk, attaquant, defenseur);
    let neutre = ResultatDegats { foudre_appliquee: val_amplifiee != val_atk, ..neutre };
    let val_atk = val_amplifiee;

    // Valeur finale du coup : la Précise double APRÈS l'amplification de la Foudre.
    let porte = if *action_atk == ActionType::P && degats_precis_doubles { val_atk * 2 } else { val_atk };
    let neutre = ResultatDegats { valeur_appliquee: porte, ..neutre };

    let chance = chance_esquive(attaquant, defenseur);
    let jet = rand::thread_rng().gen_range(1..=100);
    if jet <= chance {
        return ResultatDegats { esquive: true, degats_evites: porte, chance_esquive: chance, ..neutre };
    }

    if *action_atk == ActionType::A {
        if defenseur.armure >= porte { return ResultatDegats { dmg_arm: porte, chance_esquive: chance, ..neutre }; }
        return ResultatDegats { dmg_arm: defenseur.armure, dmg_pv: porte - defenseur.armure, chance_esquive: chance, ..neutre };
    }

    ResultatDegats { dmg_pv: porte, chance_esquive: chance, ..neutre }
}